crates/transpiler/src/standard_equivalence_library.rs

```diff
@@ -0,0 +1,3189 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::f64::consts::PI;
+use std::sync::Arc;
+
+use pyo3::prelude::*;
+use qiskit_circuit::bit::QuantumRegister;
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::operations::{Param, StandardGate};
+use qiskit_circuit::parameter::parameter_expression::ParameterExpression;
+use qiskit_circuit::parameter::symbol_expr::Symbol;
+use qiskit_circuit::Qubit;
+
+use crate::equivalence::{EquivalenceError, EquivalenceLibrary};
+
+/// Decomposition of CNOT gate.
+///
+/// NOTE: this differs to CNOT by a global phase.
+/// The matrix returned is given by exp(1j * pi/4) * CNOT
+///
+/// // Arguments:
+/// * `plus_ry` - positive initial RY rotation
+/// * `plus_rxx` - positive RXX rotation.
+///
+/// // Returns
+///
+/// * The decomposed circuit for CNOT gate (up to global phase).
+fn cnot_rxx_decompose(plus_ry: bool, plus_rxx: bool) -> PyResult<CircuitData> {
+    let sgn_ry: f64 = match plus_ry {
+        true => 1.0,
+        false => -1.0,
+    };
+    let sgn_rxx: f64 = match plus_rxx {
+        true => 1.0,
+        false => -1.0,
+    };
+    let mut circuit =
+        CircuitData::with_capacity(2, 0, 5, (-sgn_ry * sgn_rxx * PI / 4.0).into()).unwrap();
+
+    circuit.push_standard_gate(
+        StandardGate::RY,
+        &[Param::Float(sgn_ry * PI / 2.0)],
+        &[Qubit(0)],
+    )?;
+    circuit.push_standard_gate(
+        StandardGate::RXX,
+        &[Param::Float(sgn_rxx * PI / 2.0)],
+        &[Qubit(0), Qubit(1)],
+    )?;
+    circuit.push_standard_gate(
+        StandardGate::RX,
+        &[Param::Float(-sgn_rxx * PI / 2.0)],
+        &[Qubit(0)],
+    )?;
+    circuit.push_standard_gate(
+        StandardGate::RX,
+        &[Param::Float(-sgn_rxx * sgn_ry * PI / 2.0)],
+        &[Qubit(1)],
+    )?;
+    circuit.push_standard_gate(
+        StandardGate::RY,
+        &[Param::Float(-sgn_ry * PI / 2.0)],
+        &[Qubit(0)],
+    )?;
+    Ok(circuit)
+}
+
+fn create_standard_equivalence<P>(
+    gate: StandardGate,
+    params: &[Param],
+    eq_num_qubits: u32,
+    eq_instructions: &[(StandardGate, &[Qubit], &[Param])],
+    eq_global_phase: P,
+    eq_library: &mut EquivalenceLibrary,
+) -> Result<(), EquivalenceError>
+where
+    P: Into<Param>,
+{
+    let mut circuit =
+        CircuitData::with_capacity(0, 0, eq_instructions.len(), eq_global_phase.into()).unwrap();
```

Comment:
Shouldn't we preallocate the qubits here?

```suggestion
        CircuitData::with_capacity(gate.num_qubits(), 0, eq_instructions.len(), eq_global_phase.into()).unwrap();
```

Reply:
I didn't pre-allocate since we end up adding the register right after

Reply:
That doesn't change the allocation required for the qubit container types though. With capacity should preallocate the required space to store the bits information, and then adding the register populates the data in that allocated memory.

Reply:
That's true, but the way it has been implemented also adds qubits and clbits if specified. So it's not the same.

Reply:
Sigh, I didn't scroll down to the end of the method. That seems like a bug in that method, it shouldn't be doing that it goes against the normal semantics of with capacity. But we can save that for a separate PR.